### PR TITLE
Disallow opening multiple instances of an interface

### LINKED ISSUE #27112: Multiple instances of workbench interfaces should not be allowed
### Expected behavior
The reflectometry GUI should ideally be allowed to only open once, see #26981.

It has been agreed that this behaviour was standard for the cpp interfaces in MantidPlot and should be implemented into workbench for all interfaces.

### Actual behavior
Multiple instances are currently allowed

### Steps to reproduce the behavior
- Open reflectometry interface
- Open another

### Platforms affected
All

### PR DESCRIPTION
**Pull request by giovannidisiena**
**Description of work.**
This PR disallows opening multiple instances of the same interface in workbench, as was the case in MantidPlot.

**To test:**
- Open workbench
- Open any C++ interface
- Attempt to open it again
- Observe original interface raise

Fixes #27112 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/applications/workbench/workbench/app/mainwindow.py [hunk 1]
```diff
@@ -364,9 +365,14 @@ def launch_custom_python_gui(self, filename):
         self.interface_executor.execute(open(filename).read(), filename)
 
     def launch_custom_cpp_gui(self, interface_name):
-        interface = self.interface_manager.createSubWindow(interface_name)
-        interface.setAttribute(Qt.WA_DeleteOnClose, True)
-        interface.show()
+        window = find_window(interface_name, QMainWindow)
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/app/mainwindow.py]: I know you didn't add this method but could you add some method documentation to it? Especially now there is more complicated behaviour we should explain that this method creates a new window if one does not already exist else it simply shows the current window.

**Review by martyngigg — CHANGES_REQUESTED** [review]
This seems to work for most of the interfaces I tested. There is a slight oddity with "Indirect Settings". If I open workbench, open "Indirect Data Reduction" and then try to open "Indirect Settings" nothing seems to happen. If I restart workbench and try to open "Indirect Settings" first then it success.

File: qt/applications/workbench/workbench/utils/windowfinder.py [hunk 2]
```diff
@@ -25,6 +25,16 @@ def get_main_window_widget():
...
 
 
+def find_window(object_name, cls=None):
+    """
+        Finds the currently open window
+        :return: window or None
+        """
+    windows = QApplication.topLevelWidgets()
+    for window in windows:
+        if window.objectName() == object_name and isinstance(window, cls):
```
Comment (hsautessella) [inline qt/applications/workbench/workbench/utils/windowfinder.py]: I think `isinstance` will throw an error if you pass it `None` as its second argument. Perhaps make the `cls` argument required? Or probably better is to skip the `isinstance` check if no `cls` is given

**Review by ghost — DISMISSED** [review]
This worked well in my testing. Just a couple of comments

File: qt/applications/workbench/workbench/utils/windowfinder.py [hunk 3]
```diff
@@ -25,6 +25,16 @@ def get_main_window_widget():
             return widget
 
 
+def find_window(object_name, cls=None):
+    """
+        Finds the currently open window
+        :return: window or None
+        """
+    windows = QApplication.topLevelWidgets()
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply qt/applications/workbench/workbench/utils/windowfinder.py]: Good point but its more the Qt name that is a little confusing. https://doc.qt.io/qt-5/qapplication.html#topLevelWidgets says windows in brackets as I think that's what it means by top level.

File: qt/applications/workbench/workbench/utils/windowfinder.py (hunk 2 above)
Comment (martyngigg) [reply qt/applications/workbench/workbench/utils/windowfinder.py]: :+1:

**Comment by giovannidisiena** [comment]
> This seems to work for most of the interfaces I tested. There is a slight oddity with "Indirect Settings". If I open workbench, open "Indirect Data Reduction" and then try to open "Indirect Settings" nothing seems to happen. If I restart workbench and try to open "Indirect Settings" first then it success.

This is strange... I have managed to recreate this issue and it appears "Indirect Settings" is prevented from being opened after "Indirect Data Reduction", however if done first or with another interface then this is fine.

**Comment by giovannidisiena** [comment]
Indirect Settings issue should be resolved, think this is ready for re-review @martyngigg

**Review by martyngigg — CHANGES_REQUESTED** [review]

**Review by martyngigg — APPROVED** [review]
This works well now.

**Pull request merged**